### PR TITLE
Add gcp storage

### PR DESCRIPTION
**Pull request by techknowlogick**
WIP as missing docs, tests, and probably more

### CODE REVIEW

File: modules/storage/gcp.go [hunk 1]
```diff
@@ -0,0 +1,344 @@
...
+	"code.gitea.io/gitea/modules/util"
+
+	"cloud.google.com/go/storage"
+	"google.golang.org/api/iterator"
+	"google.golang.org/api/option"
+)
+
+var _ ObjectStorage = &GoogleStorage{}
+
+type googleObject struct {
```
Comment (lunny) [inline modules/storage/gcp.go]: Maybe gcpObject is a better name than googleObject? and the same to other places.

Comment (techknowlogick) [reply modules/storage/gcp.go]: thanks. I've updated this

File: modules/setting/storage.go [hunk 2]
```diff
@@ -49,12 +52,23 @@ type MinioStorageConfig struct {
...
 }
 
+// MinioStorageConfig represents the configuration for a minio storage
+type GoogleStorageConfig struct {
+	Endpoint               string `ini:"GOOGLE_ENDPOINT" json:",omitempty"`
+	ProjectID              string `ini:"GOOGLE_PROJECT_ID" json:",omitempty"`
+	Bucket                 string `ini:"GOOGLE_BUCKET" json:",omitempty"`
+	Location               string `ini:"GOOGLE_LOCATION" json:",omitempty"`
+	BasePath               string `ini:"GOOGLE_BASE_PATH" json:",omitempty"`
+	ApplicationCredentials string `ini:"GOOGLE_APPLICATION_CREDENTIALS" json:",omitempty"` // path to JSON file
```
Comment (delvh) [inline modules/setting/storage.go:62]: I haven't quite understood yet why we need the `GOOGLE_` prefix.
Shouldn't that already be obvious from the section?

File: modules/storage/gcp.go [hunk 3]
```diff
@@ -0,0 +1,344 @@
...
+func (g *gcpObject) Stat() (os.FileInfo, error) {
+	return gcpFileInfo{
+		g.Name,
+		g.Size,
+		*g.ModTime,
+	}, nil
+}
+
+// GCPStorage returns a gcp bucket storage
+type GCPStorage struct {
```
Comment (delvh) [inline modules/storage/gcp.go:102]: I think this file is weirdly structured:
You start with the assertment that the GCPStorage is a storage.
Okay.
But then I'd expect to see the definition of that type immediately afterward, with all its object methods, then the child types contained within, then the error, and only then the package methods.
Or something like that, that is at least the structure I've seen the most in Gitea so far.

File: modules/storage/gcp.go [hunk 4]
```diff
@@ -0,0 +1,344 @@
...
+}
+
+func convertGoogleErr(err error) error {
+	if err == nil {
+		return nil
+	}
+
+	// Convert two responses to standard analogues
+	// switch errResp.Code {
+	// case "NoSuchKey":
```
Comment (delvh) [inline modules/storage/gcp.go:117]: Why commented out?

File: modules/storage/gcp.go [hunk 5]
```diff
@@ -0,0 +1,344 @@
...
+func (g *gcpObject) downloadStream(p []byte) (int, error) {
+	if g.Offset > g.Size {
+		return 0, io.EOF
+	}
+	count := g.Size - g.Offset
+	pl := int64(len(p))
+	if pl > count {
+		p = p[0:count]
+	} else {
+		count = pl
```
Comment (delvh) [inline modules/storage/gcp.go:47]: count = min(count, int64(len(p)))
p = p[0:count]

File: modules/storage/gcp.go [hunk 6]
```diff
@@ -0,0 +1,344 @@
...
+
+	return n, err
+}
+
+func (g *gcpObject) Close() error {
+	return nil
+}
+
+func (g *gcpObject) Read(p []byte) (int, error) {
+	c, err := g.downloadStream(p)
```
Comment (delvh) [inline modules/storage/gcp.go:71]: return directly?

File: modules/storage/gcp.go [hunk 7]
```diff
@@ -0,0 +1,344 @@
...
+		return 0, fmt.Errorf("Seek: invalid whence")
+	case io.SeekStart:
+		offset += 0
+	case io.SeekCurrent:
+		offset += g.Offset
+	case io.SeekEnd:
+		offset += g.Size
+	}
+	if offset < 0 {
+		return 0, fmt.Errorf("Seek: invalid offset")
```
Comment (delvh) [inline modules/storage/gcp.go:87]:
```suggestion
		return 0, fmt.Errorf("Seek: invalid offset to read GCP bucket: " + offset)
```

File: modules/storage/gcp.go [hunk 8]
```diff
@@ -0,0 +1,344 @@
...
+			return false, nil
+		}
+		return false, err
+	}
+	return true, nil
+}
+
+// createBucket creates a new bucket in the specified location (region).
+func createBucket(ctx context.Context, client *storage.Client, projectID, bucketName, location string) error {
+	bkt := client.Bucket(bucketName)
```
Comment (delvh) [inline modules/storage/gcp.go:179]: `bucket :=...`

**Comment by wxiaoguang** [comment]
Would something like https://github.com/beyondstorage/go-storage help? It also helps Azure Blob Storage

**Comment by lunny** [comment]
> Would something like https://github.com/beyondstorage/go-storage help? It also helps Azure Blob Storage #25458

https://github.com/beyondstorage/go-storage looks promise.

**Comment by techknowlogick** [comment]
> https://github.com/beyondstorage/go-storage looks promise.

I'm not sure that is still actively developed. The domain was taken over by domain squatters, the only commits and PRs appear to be from dependabot.

https://gocloud.dev might be one we want to look at instead, pulumi is a project that uses it.

File: modules/setting/storage.go (hunk 2 above)
Comment (techknowlogick) [reply modules/setting/storage.go:62]: `GOOGLE_APPLICATION_CREDENTIALS` aligns with what google's libraries will look for by default, so by keeping them the same then it's less confusion for the end user. We also use `minio_` as a prefix for minio specific settings (see struct above)

**Comment by techknowlogick** [comment]
closing as gcp has s3 API support, so this is not needed.

**Comment by delvh** [comment]
@techknowlogick perhaps we should document that.
I don't think most GCP users are aware they can simply use the S3 Settings

**Pull request closed without merging**